clang-tidy-diff.sh

```diff
@@ -1,13 +1,20 @@
 #!/bin/bash
-CLANG_DIR=$(dirname $(dirname $(which clang-tidy)))
+set -e
+CLANG_TIDY=$(realpath $(which clang-tidy))
+CLANG_DIR=$(dirname $(dirname $CLANG_TIDY))
 CLANG_TIDY_DIFF=$CLANG_DIR/share/clang/clang-tidy-diff.py
-MERGE_BASE=$(git merge-base master HEAD)
-TIDY_MSG=$(git diff -U0 $MERGE_BASE | $CLANG_TIDY_DIFF -quiet -p1 2> /dev/null)
+if [ ! -f $CLANG_TIDY_DIFF ]; then
+  echo "not found: $CLANG_TIDY_DIFF"
+  exit 1
+fi
+DIFF=$(git diff -U0 "@{upstream}")
+echo "Running $CLANG_TIDY_DIFF"
+TIDY_MSG=$(echo $DIFF | $CLANG_TIDY_DIFF -quiet -p1 2> /dev/null)
 if [ -n "$TIDY_MSG" -a "$TIDY_MSG" != "No relevant changes found." ]
 then
   echo "Fix clang-tidy errors before committing!"
   echo
   # Run clang-tidy once again to show the error
-  git diff -U0 $MERGE_BASE | $CLANG_TIDY_DIFF -quiet -p1 2> /dev/null
+  git diff -U0 "@{upstream}" | $CLANG_TIDY_DIFF -quiet -p1 2> /dev/null
```

Comment:
What does `"@{upstream}" mean? Does this work when we the current PR is based on an older commit? For example, if the master branch commits are like
```
0 -> 1 -> 2 -> 3
                        (master is at 3)
```

Currently the master is at 3 but a PR was created when the master pointed to the commit 2. So in this case we we shouldn't take a diff between the current master and the PR. Does this work in this case too?

Reply:
`@{upstream}` is just whatever the upstream of the current branch is set to.

We could switch to `origin/master` if you like.  But I don't think a local branch called `master` exists so I don't see how this could have ever worked.  At least it I don't have a local branch called `master` .. maybe travis does?

Reply:
Sorry, so the upstream means master here? Or do all repos have something called upstream branch?

And the other thing I asked about is `git merge-base` in the current file. That takes the common base ancestor of master branch and the PR branch. Even if we change master to upstream, shouldn't we still do the `git merge-base`?

Reply:
Oh, does 'upstream' mean the branch a PR tries to merge into? Then that it makes sense... Anyway `git merge-base` looks necessary in that case too